### PR TITLE
Add (back in) GCP PubSub StreamProvider

### LINKED ISSUE #7509: Upgrade Google.PubSub

### LINKED ISSUE #8264: Remove Google PubSub support
See #7509

Our Google PubSub library has been untouched for a long time, relies on an out-dated prerelease client, and I assume it is essentially unused. It's likely better to remove it and resurrect it in the event that someone wants to maintain it.

### PR DESCRIPTION
**Pull request by abbottdev**
Fixes #7509
Closes #8264

I'd like to start a conversation about what other work needs doing to get GCP PubSub Streams included back into Orleans. 

I've got the compilation errors resolved, but this is my first time contributing to this repo so any tasks/advice I'm more than happy to have a go at.
 

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/orleans/pull/8572)

### CODE REVIEW

**Comment by ReubenBond** [comment]
Hi @abbottdev, thank you for opening the PR! We need to get the tests working in GitHub Actions, ideally using an emulator or some other containerized service. Do you have much familiarity with that?

File: src/Orleans.Streaming.GCP/Orleans.Streaming.GCP.csproj [hunk 1]
```diff
@@ -11,7 +11,7 @@
 
   <PropertyGroup>
     <!-- Cannot release a stable version of this package that depends on a prerelease version of Google.Cloud.PubSub.V1. -->
-    <IsPackable>false</IsPackable>
+    <IsPackable>true</IsPackable>
```
Comment (ReubenBond) [inline src/Orleans.Streaming.GCP/Orleans.Streaming.GCP.csproj:14]: Maybe we should rename this package to `Microsoft.Orleans.Streaming.GoogleCloudPubSub` to match the dependency package name (i.e, Google.Cloud.PubSub.V1)

**Comment by ReubenBond** [comment]
The main thing this needs is test automation in GitHub Actions. Let me know if you would like some help with that. The typical approach is to use a containerized version of the service or an emulator (see the [existing jobs](https://github.com/dotnet/orleans/blob/82f3e88200d2dfe9df4429378ab52c97639e3b86/.github/workflows/ci.yml#L36) for example)

**Comment by abbottdev** [comment]
@ReubenBond I've had a go at the action for now - Happy to update the namespace if required. The tests run locally as follows:

1. Start gcloud pubsub emulator via `gcloud beta emulators pubsub start --project=test-proj`
2. Set `PUBSUB_EMULATOR_HOST` to `http://localhost:8085` (This also relaxes SSL restrictions on the PubSub libraries)
3. `dotnet test`

Given the above, I've tried to setup ci.yaml to use a dockerized version of that emulator - but on a different port. 

I'm seeing some test failures, but I can't tell at this point whether this is because the Orleans Streaming API has moved on and there are gaps in the implementation, or, whether there's something fundamentally wrong with the tests.

**Comment by ReubenBond** [comment]
I see that the emulator is still failing for some reason in GitHub

**Comment by abbottdev** [comment]
@ReubenBond apologies for the delay on this one - Looks like a missing env var on my part. Could we get another run?

**Comment by ReubenBond** [comment]
Sure thing, @abbottdev - I've rerun them

**Comment by jeremylcarter** [comment]
@ReubenBond 

Can you rerun these tests so that I can get a copy of the logs and error messages?

Thanks!

**Comment by ReubenBond** [comment]
@jeremylcarter running them now

**Comment by abbottdev** [comment]
@ReubenBond  @jeremylcarter  sorry I've been AWOL - I've pushed up some changes to make the emulator use the gcloud emulator with the correct environment variables set (🤞 ). 

Could you kick off another run? I was seeing some parallism test failures when I ran the tests locally but I was a bit stumped where to start with them.

**Comment by ReubenBond** [comment]
I kicked off a run

**Comment by abbottdev** [comment]
Build result:
```
gcr.io/google.com/cloudsdktool/google-cloud-cli:emulators
  /usr/bin/docker create --name 743cedb95dd645fb996285a3[53](https://github.com/dotnet/orleans/actions/runs/7421651788/job/20202172711#step:2:56)ce1f47_gcriogooglecomcloudsdktoolgooglecloudcliemulators_f120b2 --label d6736e --network github_network_01e8706b6b0d41e9b1ea5bd895f31027 --network-alias emulator -p 8085:8085 --entrypoint gcloud beta emulators pubsub start --project=feisty-flow-173313 --host-port=0.0.0.0:8085 -e GITHUB_ACTIONS=true -e CI=true gcr.io/google.com/cloudsdktool/google-cloud-cli:emulators
  Unable to find image 'beta:latest' locally
  Error response from daemon: pull access denied for beta, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
  Error: Exit code 1 returned from process: file name '/usr/bin/docker', arguments 'create --name 743cedb95dd645fb996285a353ce1f47_gcriogooglecomcloudsdktoolgooglecloudcliemulators_f120b2 --label d6736e --network github_network_01e8706b6b0d41e9b1ea5bd895f31027 --network-alias emulator -p 8085:8085 --entrypoint gcloud beta emulators pubsub start --project=feisty-flow-173313 --host-port=0.0.0.0:8085 -e GITHUB_ACTIONS=true -e CI=true gcr.io/google.com/cloudsdktool/google-cloud-cli:emulators'.
```

The above looks like the `--entrypoint` command needs to be quoted - which I've just pushed up a change for. 🤞 that's enough. @ReubenBond sorry to have to keep asking, but could we kick off another build?

**Comment by ReubenBond** [comment]
Closing for now, since GCP PubSub was removed by #9031

**Comment by jeremylcarter** [comment]
I initially had a go at this a year ago. I no longer need GCP as we've migrated fully to Azure. Anything I can do to help?

**Pull request closed without merging**